File: kakuro.py
import random

# algorithme de backtracking pour génération de kakuro

class grilleKakuro:
    def __init__(self, n, m, difficulte):
        self.n = n
        self.m = m
        self.maxi = max(n, m)
        self.grille = [[0 for i in range(m)] for j in range(n)]
        self.entete = [[0 for i in range(m)] for j in range(n)]
        self.difficulte = difficulte
        self.genererGrille()

    # on remplit la grille avec des des valeurs de 1 à maxi
    def genererGrille(self):
        self.genererTrou()
        # for i in range(self.n):
        #     for j in range(self.m):
        #         if self.grille[i][j] == 0:
        #             valeur = random.randint(1, self.maxi)
        #             while self.estDansLigneOuColonne(i, j, valeur):
        #                 valeur = random.randint(1, self.maxi)
        #             self.grille[i][j] = valeur
        # self.estValide(0, 0, 0)
        self.estValideRandom(0, 0, 0)
        self.genererEntete()
        self.deleteUselessTuples()
        

    def estDansLigneOuColonne(self, i, j, val):
        # on vérifie que la valeur n'est pas déjà dans la ligne ou la colonne, on s'arrête si on est sur un trou
        k = 0
        while k < self.m and self.grille[i][k] != -1:
            if self.grille[i][k] == val:
                return True
            k += 1
        k = 0
        while k < self.n and self.grille[k][j] != -1:
            if self.grille[k][j] == val:
                return True
            k += 1
        return False
    
    # algo de backtracking
    def estValide(self, i, j, position):
        if position == self.n * self.m:
            return True
        
        if self.grille[i][j] != 0:
            if j == self.m - 1:
                return self.estValide(i + 1, 0, position + 1)
            else:
                return self.estValide(i, j + 1, position + 1)
            
        for val in range(1, self.maxi + 1):
            if not self.estDansLigneOuColonne(i, j, val):
                self.grille[i][j] = val
                if j == self.m - 1:
                    if self.estValide(i + 1, 0, position + 1):
                        return True
                else:
                    if self.estValide(i, j + 1, position + 1):
                        return True
                self.grille[i][j] = 0
        return False
    
    # on refait la même fonction en utilisant des valeurs aléatoires pour la génération de la grille
    def estValideRandom(self, i, j, position):
        if position == self.n * self.m:
            return True
        
        if self.grille[i][j] != 0:
            if j == self.m - 1:
                return self.estValideRandom(i + 1, 0, position + 1)
            else:
                return self.estValideRandom(i, j + 1, position + 1)
            
        for val in range(1, self.maxi + 1):
            valeur = random.randint(1, self.maxi)
            if not self.estDansLigneOuColonne(i, j, valeur):
                self.grille[i][j] = valeur
                if j == self.m - 1:
                    if self.estValideRandom(i + 1, 0, position + 1):
                        return True
                else:
                    if self.estValideRandom(i, j + 1, position + 1):
                        return True
                self.grille[i][j] = 0
        return False
    
    
    # on fait des trous au hasard dans la grille en fonction de la difficulté, plus la difficulté est grande, moins il y a de trous
    def genererTrou(self):
        for i in range(self.n):
            for j in range(self.m):
                if random.randint(1, self.difficulte) == 1:
                    self.grille[i][j] = -1

        self.checkTrou()
        self.genererEntete()

    # on vérifie si chaque ligne et chaque colonne n'a pas plus de 9 cases d'affilée
    # si c'est le cas, on remplace une valeur par un trou aléatoirement
    def checkTrou(self):
        for i in range(self.n):
            for j in range(self.m):
                if self.grille[i][j] != -1:
                    k = j
                    while k < self.m and self.grille[i][k] != -1:
                        k += 1
                    if k - j > 9:
                        self.grille[i][random.randint(j, k - 1)] = -1
        for j in range(self.m):
            for i in range(self.n):
                if self.grille[i][j] != -1:
                    k = i
                    while k < self.n and self.grille[k][j] != -1:
                        k += 1
                    if k - i > 9:
                        self.grille[random.randint(i, k - 1)][j] = -1

    # on génère les entêtes de la grille
    def genererEntete(self):
        # On doit avoir un tuple pour chaque case de la grille indiquant la somme de la colonne et de la ligne
        # si la case est un trou, on met un tuple vide
        for i in range(self.n):
            for j in range(self.m):
                if self.grille[i][j] == -1:
                    self.entete[i][j] = ()
                else:
                    self.entete[i][j] = (self.sommeLigne(i, j), self.sommeColonne(i, j))


    # on calcule la somme de la ligne
    def sommeLigne(self, i, j):
        # on s'arrête si on est sur un trou et on retourne la somme
        somme = 0
        while j < self.m and self.grille[i][j] != -1:
            somme += self.grille[i][j]
            j += 1
        return somme
    
    # on calcule la somme de la colonne
    def sommeColonne(self, i, j):
        # on s'arrête si on est sur un trou et on retourne la somme
        somme = 0
        while i < self.n and self.grille[i][j] != -1:
            somme += self.grille[i][j]
            i += 1
        return somme
    
    # on affiche la grille
    def affiche(self):
        for i in range(self.n):
            for j in range(self.m):
                print(self.grille[i][j], end = " ")
            print()

    # on affiche les entêtes
    def afficheEntete(self):
        for i in range(self.n):
            for j in range(self.m):
                print(self.entete[i][j], end = " ")
            print()

    def deleteUselessTuples(self):
        # on met à 0 le premier tuple si on a un voisin à gauche
        for i in range(self.n):
            for j in range(self.m):
                if self.grille[i][j] != -1 and j > 0 and self.grille[i][j - 1] != -1:
                    self.entete[i][j] = (0, self.entete[i][j][1])
        # on met à 0 le premier tuple si on a un voisin en haut
        for i in range(self.n):
            for j in range(self.m):
                if self.grille[i][j] != -1 and i > 0 and self.grille[i - 1][j] != -1:
                    self.entete[i][j] = (self.entete[i][j][0], 0)
        # change all the tuples with o,0 to an empty tuple
        for i in range(self.n):
            for j in range(self.m):
                if self.entete[i][j] == (0, 0):
                    self.entete[i][j] = ()

    def enteteVide(self, i, j):
        return self.entete[i][j] == ()

    def getGrille(self):
        return self.grille
    
    
test = grilleKakuro(6, 6, 10)
test.affiche()
print()
test.afficheEntete()

File: kakuro.dart
import 'dart:math';

class Kakuro {
  late int n, m, maxi, difficulte;
  late dynamic grille, entete;

  Kakuro(int n, int m, int difficulte) {
    grille = List.generate(n, (i) => List.generate(m, (j) => 0));
    entete = List.generate(n, (i) => List.generate(m, (j) => 0));
    maxi = max(n, m);
    this.n = n;
    this.m = m;
    this.difficulte = difficulte;

    genererGrille();
  }

  void genererGrille() {
    genererTrou();
    estValideRandom(0, 0, 0);
    genererEntete();
    deleteUselessTuples();
  }

  bool estDansLigneOuColonne(int i, int j, int val) {
    for (int k = 0; k < m; k++) {
      if (grille[i][k] == val) {
        return true;
      }
    }
    for (int k = 0; k < n; k++) {
      if (grille[k][j] == val) {
        return true;
      }
    }
    return false;
  }

  void genererTrou() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (Random().nextInt(difficulte).toInt() == 1) {
          grille[i][j] = -1;
        }
      }
    }
    checkTrou();
    genererEntete();
  }

  void checkTrou() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (grille[i][j] == -1) {
          int k = j;
          while (k < m && grille[i][k] != -1) {
            k++;
          }
          if (k - j > 9) {
            grille[i][Random().nextInt(k - j) + j] = -1;
          }
        }
      }
    }
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (grille[j][i] == -1) {
          int k = j;
          while (k < n && grille[k][i] != -1) {
            k++;
          }
          if (k - j > 9) {
            grille[Random().nextInt(k - j) + j][i] = -1;
          }
        }
      }
    }
  }

  void genererEntete() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (grille[i][j] == -1) {
          entete[i][j] = [];
        } else {
          entete[i][j] = [sommeLigne(i, j), sommeColonne(i, j)];
        }
      }
    }
  }

  int sommeLigne(int i, int j) {
    num somme = 0;
    int k = j;
    while (k < m && grille[i][k] != -1) {
      somme += grille[i][k];
      k++;
    }
    return somme.toInt();
  }

  int sommeColonne(int i, int j) {
    num somme = 0;
    int k = i;
    while (k < n && grille[k][j] != -1) {
      somme += grille[k][j];
      k++;
    }
    return somme.toInt();
  }

  void deleteUselessTuples() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (grille[i][j] != -1 && j > 0 && grille[i][j - 1] != -1) {
          entete[i][j] = [0, entete[i][j][1]];
        }
        if (grille[i][j] != -1 && i > 0 && grille[i - 1][j] != -1) {
          entete[i][j] = [entete[i][j][0], 0];
        }
        if (entete[i][j] == [0, 0]) {
          entete[i][j] = [];
        }
      }
    }
  }

  bool estValide(i, j, position) {
    if (position == n * m) return true;
    if (grille[i][j] != 0) {
      if (j == m - 1) {
        return estValide(i + 1, 0, position + 1);
      } else {
        return estValide(i, j + 1, position + 1);
      }
    }

    for (int val = 1; val <= maxi; val++) {
      if (!estDansLigneOuColonne(i, j, val)) {
        grille[i][j] = val;
        if (j == m - 1) {
          if (estValide(i + 1, 0, position + 1)) {
            return true;
          }
        } else {
          if (estValide(i, j + 1, position + 1)) {
            return true;
          }
        }
        grille[i][j] = 0;
      }
    }
    return false;
  }

  bool estValideRandom(i, j, position) {
    if (position == n * m) return true;
    if (grille[i][j] != 0) {
      if (j == m - 1) {
        return estValideRandom(i + 1, 0, position + 1);
      } else {
        return estValideRandom(i, j + 1, position + 1);
      }
    }
    for (int val = 1; val <= maxi; val++) {
      int valeur = Random().nextInt(maxi) + 1;
      if (!estDansLigneOuColonne(i, j, valeur)) {
        grille[i][j] = valeur;
        if (j == m - 1) {
          if (estValideRandom(i + 1, 0, position + 1)) {
            return true;
          }
        } else {
          if (estValideRandom(i, j + 1, position + 1)) {
            return true;
          }
        }
        grille[i][j] = 0;
      }
    }
    return false;
  }

  void affiche() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        print(grille[i][j]);
      }
      print("");
    }
  }

  void afficheEntete() {
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        print(entete[i][j]);
      }
      print("");
    }
  }
}

// test main
void main() {
  Kakuro g = Kakuro(5, 5, 9);
  g.affiche();
  g.afficheEntete();
}
